### PR TITLE
Enable compiling with crossgen2

### PR DESCRIPTION
**Pull request by davidwrighton**
Enable using crossgen2 instead of crossgen in build. Building block for enabling pgo for this repo.

Addresses #31040 (in this specific format)

### CODE REVIEW

**Review by dougbu — COMMENTED** [review]
Lemme know when CI build looks greener and mark the PR as ready for review at the same point please

File: eng/Dependencies.props [hunk 1]
```diff
@@ -206,7 +214,7 @@ and are generated based on the last package release.
     -->
     <LatestPackageReference Update="@(LatestPackageReference->WithMetadataValue('Version', ''))">
       <Version
-          Condition=" $([System.String]::new('%(Identity)').StartsWith('Microsoft.NETCore.App.Runtime.')) ">$(MicrosoftNETCoreAppRuntimeVersion)</Version>
+          Condition=" $([System.String]::new('%(Identity)').StartsWith('Microsoft.NETCore.App.Runtime.')) or $([System.String]::new('%(Identity)').StartsWith('Microsoft.NETCore.App.Crossgen2.'))">$(MicrosoftNETCoreAppRuntimeVersion)</Version>
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply eng/Dependencies.props:217]: Yes. They should always be exactly aligned.

**Comment by davidwrighton** [comment]
@dougbu Can you take a look? I also attempted to add some tests of the generated product by removing the disabling of crossgen2 in a spot in the official build, but I don't have confidence this actually did anything useful. My plan is to remove that change to the yaml before checkin.

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 2]
```diff
@@ -125,6 +133,11 @@ This package is an internal implementation of the .NET Core SDK and is not meant
         PrivateAssets="All"
         GeneratePathProperty="true" />
 
+    <Reference Include="Microsoft.NETCore.App.Crossgen2.$(BuildOsName)-$(BuildArchitecture)"
+        ExcludeAssets="All"
+        PrivateAssets="All"
+        GeneratePathProperty="true" />
```
[the remark this answers is not in the document]
Comment (dougbu) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:139]: Any possibility of removing the need for workarounds (hacks) like https://github.com/dotnet/aspnetcore/pull/31778/files#diff-eb8fd31a7971d712606a187ea074d0bd5d6ea85aae5a707107fd998216986ea5R117, https://github.com/dotnet/aspnetcore/pull/31778/files#diff-eb8fd31a7971d712606a187ea074d0bd5d6ea85aae5a707107fd998216986ea5R365-R368, and related bits in the future❔

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj [hunk 3]
```diff
@@ -413,18 +446,35 @@ This package is an internal implementation of the .NET Core SDK and is not meant
...
           Condition="'$(CrossCompileDirectory)' != ''">$(RuntimePackageRoot)runtimes\$(CrossCompileDirectory)\native\$(LibPrefix)clrjit$(LibExtension)</CoreCLRJitPath>
+
+      <Crossgen2TargetOs>$(TargetOsName)</Crossgen2TargetOs>
+      <Crossgen2TargetOs Condition="'$(TargetOsName)' == 'win'">windows</Crossgen2TargetOs>
+      <Crossgen2TargetOs Condition="'$(TargetOsName)' == 'linux-musl'">linux</Crossgen2TargetOs>
+      <Crossgen2Args>--targetarch:$(TargetArchitecture)</Crossgen2Args>
+      <Crossgen2Args>$(Crossgen2Args) --targetos:$(Crossgen2TargetOs)</Crossgen2Args>
+      <Crossgen2Args>$(Crossgen2Args) -O</Crossgen2Args>
+      <Crossgen2Args>$(Crossgen2Args) @&quot;$(CrossgenToolDir)PlatformAssembliesPathsCrossgen2.rsp&quot;</Crossgen2Args>
+      <Crossgen2Args Condition="Exists('$(RuntimePackageRoot)tools\StandardOptimizationData.mibc')">$(Crossgen2Args) &quot;-m:$(RuntimePackageRoot)tools\StandardOptimizationData.mibc&quot;</Crossgen2Args>
```
Comment (dougbu) [inline src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:457]: What is the StandardOptimizationData.mibc file and does it have anything to do with ASP.NET Core scenarios❔ I as expecting to check in PGO data (if that's what this file is about) specific to this repo.

File: .azure/pipelines/quarantined-pr.yml [hunk 4]
```diff
@@ -65,7 +65,7 @@ jobs:
     - script: ./eng/build.cmd -ci -nobl -noBuildRepoTasks -noRestore -test -all -noBuildJava -noBuildNative
               -projects eng\helix\helix.proj /p:RunQuarantinedTests=true /p:IsRequiredCheck=true /p:IsHelixJob=true
               /p:BuildInteropProjects=true /p:RunTemplateTests=true
-              /p:CrossgenOutput=false /p:ASPNETCORE_TEST_LOG_DIR=artifacts/log
+              /p:ASPNETCORE_TEST_LOG_DIR=artifacts/log
```
Comment (dougbu) [inline .azure/pipelines/quarantined-pr.yml:68]: This isn't a particularly interesting change given we _expect_ failures in the quarantined runs. Suggest instead removing `/p:CrossgenOutput=false` from the 'Tests: Helix x64' job's build steps in ci.yml. If the `crossgen2`'d assemblies are platform-specific with the command arguments you're currently using, I'd expect to see a lot of failures on Linux and macOS work items (ProjectTemplates.Test in particular) submitted from that job without the property setting.

**Comment by dougbu** [comment]
> My plan is to remove that change to the yaml before checkin.

Reverting that change makes sense but I'd like to see test failures showing we're producing the best images for each platform i.e. that we hit problems when `crossgen2` is enabled.

Note we only use the shared Fx assemblies for a limited number of tests. Search for `<TestDependsOnAspNetPackages>true</TestDependsOnAspNetPackages>` in Helix-enabled test projects to find them. Might be worth copying Microsoft.AspNetCore.App.Runtime files over the assemblies that land in the Helix work item root to get unit tests using the `crossgen2`'d assemblies.

File: src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj (hunk 3 above)
Comment (davidwrighton) [reply src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj:457]: StandardOptimizationData is the training that is produced from the optimization repo. It involves a variety of scenarios, such as TechEmpower plaintext mvc, a simple helloworld, various sdk activities. There are plans to expand the coverage of asp.net scenarios, by adding test scenarios such as Orchard as well. The end result is that StandardOptimizationData holds optimization data relevant to this repo. If you wish to build up a parallel infrastructure, it is in theory possible, but my recommendation is to just piggy back on this data.

**Comment by davidwrighton** [comment]
Doug is taking over this with #32508,  so I'm closing this PR.

**Pull request closed without merging**